Added several new types of portal environment

Night-Pryanik opened the pull request:
#### Summary
```SUMMARY: Content "Added several new types of portal environment."```

#### Purpose of change
Portal (the "tear in reality" type) always seemed kinda boring to me, as well as inconsistent - it spawned only non-fungus netherworld creatures and at the same time it spawned fungal terrain. So I decided to add some new content for this kind of stuff and tweak existing content.

#### Describe the solution
Additions and changes to portal environment:
- fungal (ordinary and marloss) terrains with a chance to spawn a pouf-makers;
- netherworld creatures (almost no changes to the pre-this-PR-version, except for removed fungal terrain);
- several cracks in the ground originating from the portal and caused by it;
- radiation from the portal killed all the vegetation in some radius;
- lava seams originating from the portal (I used existing fumaroles mapgen and applied it to the portal mapgen. Fumaroles won't be spawning in random places around the world for no apparent reason anymore. They will spawn only at the portals);
- mi-go scouts went through the portal and began constructing their base of operations;
- portal created an anomaly which in turn spawned an artifact (so from now on anomalies will be created only as a part of portals environment, and won't be spawning by themselves).

#### Additional context
Fungal version
![11111111](https://user-images.githubusercontent.com/11132525/61744758-1da5e380-ada9-11e9-812c-0e9b5e254f6d.png)

Netherworld creatures version
![22222222](https://user-images.githubusercontent.com/11132525/61744816-3a421b80-ada9-11e9-8b26-0e45e872352f.png)

Failures version
![33333333](https://user-images.githubusercontent.com/11132525/61744831-4201c000-ada9-11e9-940f-8af1a719d3ba.png)

Radiation and dead vegetation version
![444444444](https://user-images.githubusercontent.com/11132525/61744859-4e861880-ada9-11e9-9e33-605d3ee39f4d.png)

Fumarole version
![555555555](https://user-images.githubusercontent.com/11132525/61744867-55ad2680-ada9-11e9-94d5-87cb5cf7a313.png)

Mi-go version
![изображение](https://user-images.githubusercontent.com/11132525/61768612-042c8800-adf9-11e9-8dde-87e7852d5d31.png)

BevapDin commented on src/map_extras.cpp:
```diff
@@ -100,6 +102,58 @@ static void mx_null( map &, const tripoint & )
+    // terrain specific conversions
+    const ter_id tid = m.ter( loc );
+    static const std::map<ter_id, ter_str_id> dies_into {{
+            {t_grass, ter_str_id( "t_grass_dead" )},
```
Being a static variable means it's loaded only once, but there is no guarantee the integer ids are the same upon the next run. Image the player saves their game (after encountering this) and loads another game with other mods that change the integer ids of the terrains here.

I suggest using `ter_str_id` instead of `ter_id`.

This also looks like a good usage of the infrastructure introduced by #32329.

Night-Pryanik replied:
That's not my code, I simply moved all code snippet to the very beginning of the file.

BevapDin commented on src/map_extras.cpp:
```diff
@@ -1399,17 +1453,130 @@ static void mx_fumarole( map &m, const tripoint &abs_sub )
+        //Netherworld monsters spawning around the portal
+        case 2: {
+            m.add_field( portal_location, fd_fatigue, 3 );
+            for( int i = x - 5; i <= x + 5; i++ ) {
```
Why are you using an explicit iteration over x,y here instead of the `points_in_radius` like above?

Night-Pryanik replied:
It is a left-over from a previous implementation, I simply didn't change it.

Night-Pryanik replied:
Fixed.

BevapDin commented on src/map_extras.cpp:
```diff
@@ -1399,17 +1453,130 @@ static void mx_fumarole( map &m, const tripoint &abs_sub )
+
+                // Pick a random cardinal direction to also spawn lava in
+                // This will make the lava a single connected line, not just on diagonals
+                std::vector<direction> possibilities;
```
```suggestion
                static const std::array<direction, 4> possibilities = { { EAST, WEST, NORTH, SOUTH } };
```

kasanryukin replied:
Don't we have 8 possible directions? Vector makes more sense.

BevapDin replied:
> Vector makes more sense.

Why? This is never changed at runtime and the number of elements is known at compile-time, so it might as well be an `array`. Note: this is `std::vector`, not the geometric vector.

> Don't we have 8 possible directions?

See comment above that code: it picks a *cardinal direction*.

kasanryukin replied:
Ooooh I see.  Reading is fundamental.  And requires coffee.

Night-Pryanik replied:
Fixed.

BevapDin commented on src/map_extras.cpp:
```diff
@@ -1399,17 +1453,130 @@ static void mx_fumarole( map &m, const tripoint &abs_sub )
+
+                for( auto &i : ignited ) {
+                    // Don't need to do anything to tiles that already have lava on them
+                    if( m.ter( i.x, i.y ) != t_lava ) {
```
```suggestion
                    if( m.ter( i ) != t_lava ) {
```

BevapDin commented on src/map_extras.cpp:
```diff
@@ -1399,17 +1453,130 @@ static void mx_fumarole( map &m, const tripoint &abs_sub )
+                for( int j = y - 5; j <= y + 5; j++ ) {
+                    if( rng( 1, 9 ) >= trig_dist( x, y, i, j ) ) {
+                        if( one_in( 15 ) ) {
+                            m.place_spawns( GROUP_MI_GO_CAMP_OM, 1, i, j, i, j, 1, true );
```
Why not leave the placing to `place_spawns`? It allows specifying and area instead of a single point. This would also ensure the creatures don't have to spawn on impassable terrain.

Something like:
```suggestion
                            m.place_spawns( GROUP_MI_GO_CAMP_OM, 1, x-5, y-5, x+5, y+5, 1, true );
```
(You may have to increase the chance / density.)

Night-Pryanik replied:
Fixed.

anothersimulacrum commented on src/map_extras.cpp:
```diff
@@ -1399,17 +1454,121 @@ static void mx_fumarole( map &m, const tripoint &abs_sub )
+    const int x = portal_location.x;
+    const int y = portal_location.y;
+
+    switch( rng( 6, 6 ) ) {
```
```suggestion
    switch( rng( 1, 6 ) ) {
```
Unless you only want to Mi-Go one, of course

Night-Pryanik replied:
That's a debug random number generator 😄

Fris0uman commented:
Aren't Mi-go from space in our reality and not from the nether?

Night-Pryanik commented:
> Aren't Mi-go from space in our reality and not from the nether?

Not sure about that, but even if they are, tears in reality not necessarily connected only to the nether. They are portals, after all, and could be connected to far-away places in our cosmos.

Night-Pryanik commented:
Added seventh variant in which I merged anomaly map extra. So in 7th variant portal will spawn anomaly and artifacts. I removed anomaly map extra from `regional_settings.json` and increased chance to spawn portal map extra accordingly. Is it acceptable?

Inglonias commented on src/map_extras.cpp:
```diff
@@ -1399,21 +1454,130 @@ static void mx_fumarole( map &m, const tripoint &abs_sub )
+        //Mycus spreading through the portal
+        case 1: {
+            m.add_field( portal_location, fd_fatigue, 3 );
+            fungal_effects fe( *g, m );
```
Does this account for the no fungals mod?

Night-Pryanik replied:
No, and it wasn't accounted for in the previous version.
No fungal mod only blacklists fungal monsters, overmap specials, and removes iexamine functions from two types of terrain. There is no way to blacklist hardcoded fungalization by the means of mod definition.
